### ISSUE TITLE
Error when spinning up a child process from inside of ruby

### ISSUE THREAD

**Issue by david22swan** [puppetlabs/pdk#1083]
When running the Puppet.Dsc acceptance tests on GHA we hit an error during the setup for the first test.
This error occurs whenever the pdk is called upon in order to create a module and states that there has been a `ChildProcess::LaunchError` due to our access having been denied. 

The exact error that we have been getting can be seen [here](https://github.com/puppetlabs/Puppet.Dsc/runs/2730261890?check_suite_focus=true#step:6:44).
And the PR on which we were working when we discovered this bug can be shown [here](https://github.com/puppetlabs/Puppet.Dsc/pull/159).

A conversation between @michaeltlombardi  the inimitable @glennsarti, who went out of his way to try a minimal local repro using the GHA docker image, discovered the problem seems to be this line in the PDK, which sets `process.leader`:

https://github.com/puppetlabs/pdk/blob/2b03b72ed71be20bf77b415b5709265164ce88dc/lib/pdk/cli/exec/command.rb#L30

In his manual testing, attempting to start a process in GHA with this flag breaks

```ruby
p = ChildProcess.build('ls')
p.leader = true
p.start
p.wait
puts p.exit_code
# Access is denied
```

But with it removed, no error is returned:

```ruby
p = ChildProcess.build('ls')
p.start
p.wait
puts p.exit_code
```

We had opened a PR on the pdk in order to try and implement two possible solutions, shown [here](https://github.com/puppetlabs/pdk/pull/1081).

The first of these solutions was to move the pin of the `childprocess` gem from `v0.7.1` to `v4.0.0` in order to bring in the patch shown [here](https://github.com/enkessler/childprocess/pull/167/commits), however this did not resolve the issue as the error still occured.

The second of our attempted solutions was to comment out the above linked line:
https://github.com/puppetlabs/pdk/blob/2b03b72ed71be20bf77b415b5709265164ce88dc/lib/pdk/cli/exec/command.rb#L30

But this did not work either as the change somehow caused a dependency error in the bundle install as shown [here](https://github.com/puppetlabs/Puppet.Dsc/runs/2737918820?check_suite_focus=true#step:6:399)

A draft for the experimental branch that we used is currently open can can be found here: #1081 

We are hoping for some help in finding a fix for this as it is a major blocker in moving our CI off of appveyor.

**Comment by jpogran** [comment]
Attempted to run locally, works fine.
Attempted to run in a windows container, works fine.
Attempted to run in a bolt remote command, works fine.

Did a little research, it seems likes others are having similar problems in Github: https://github.com/enkessler/childprocess/pull/178/checks?check_run_id=2725513511

**Comment by sanfrancrisko** [comment]
Thanks to @da-ar for the investigation in to this, so far.

We use the [childprocess](https://github.com/enkessler/childprocess) to abstract the spawning / handling of exec threads across platforms. Setting the [leader](https://github.com/enkessler/childprocess/blob/c8145d8cc9832939a5926ee70d5162749b56a958/README.md#ensure-entire-process-tree-dies) attribute will ensure any process trees created will die when the child process does.

On Windows platforms, this is done by [setting the CREATE_BREAKAWAY_FROM_JOB flag](https://github.com/enkessler/childprocess/blob/c6260739e86c1cabd9baebd76f2058eea2b1ae16/lib/childprocess/windows/process_builder.rb#L105) when calling [AssignProcessToJobObject](https://docs.microsoft.com/en-us/windows/win32/api/jobapi2/nf-jobapi2-assignprocesstojobobject] in the Win32 API. According to [the Win32 API documentation]https://docs.microsoft.com/en-us/windows/win32/api/jobapi2/) when this flag is set:

> The child process breaks away from every job in the job chain unless a job in the chain does not allow breakaway. In this case, the child process does not break away from that job or any job above it in the job chain

Jobs (and nested jobs) created through this API are created via the Task Scheduler. According to [this bug write up](https://bugzilla.mozilla.org/show_bug.cgi?id=1330718) from Mozilla, and [the associated patch to fix](https://github.com/taskcluster/runlib/pull/2) on Windows < 8 / Server 2008 R2 ( NT < 6.2 ), the `CREATE_BREAKAWAY_FROM_JOB` flag required `JOB_OBJECT_LIMIT_BREAKAWAY_OK` flag to be set too (which `childprocess` does [here](https://github.com/enkessler/childprocess/blob/c6260739e86c1cabd9baebd76f2058eea2b1ae16/lib/childprocess/windows/process.rb#L105), however, on Windows > Windows 8 / Server 2008 R2, the Win32 API will return an ACCESS_DENIED error if the `JOB_OBJECT_LIMIT_BREAKAWAY_OK` flag is set. On a Windows 10 system it was observed that Task Scheduler did not set this flag as nested jobs are supported.

For our scenario, we should conditionally set the [`@process.leader`](https://github.com/puppetlabs/pdk/blob/6b1aec6f819fcf154e1417d8ef6a4340578b749a/lib/pdk/cli/exec/command.rb#L30) parameter to true if the OS is <= Windows 7 / Server 2003, and in other cases, leave it unset / explicitly set to false.

More investigation / understanding would be required to offer a fix upstream to [`enkessler/childprocess`](https://github.com/enkessler/childprocess), as there's a lot of unknowns about how how these jobs are created (i.e. where Task Scheduler fits in, how the process / child processes are run - logon service vs user) that I'm simply not comfortable enough with at the moment.

**Comment by sanfrancrisko** [comment]
Bumping [`childprocess` to version `4.0.0`](https://github.com/sanfrancrisko/pdk/commit/2ee09b8cf230ae6a1cbd36bf0894eb4e956d36d0) and [ensuring `@process.leader` is unset](https://github.com/sanfrancrisko/pdk/commit/0febbcb9a199242285abf6c371afd8797da42d13) has seemed to resolve the issue on Windows Server 2016 and 2019:

Using [this test workflow](https://github.com/sanfrancrisko/puppetlabs-testing/pull/2/files#diff-46b15f1dd2f38b79484e37414eb1d18760dc72eec4e62acd5dd8341d7c9cf03a), on a repo [pulling in a version of PDK with the above changes](https://github.com/sanfrancrisko/puppetlabs-testing/pull/2/files#diff-d09ea66f8227784ff4393d88a19836f321c915ae10031d16c93d67e6283ab55f), I was able to get `bundle exec pdk validate` to run successfully:

- [Windows Server 2019](https://github.com/sanfrancrisko/puppetlabs-testing/actions/runs/918344364)
- [Windows Server 2016](https://github.com/sanfrancrisko/puppetlabs-testing/runs/2783640338)

[This run](https://github.com/sanfrancrisko/puppetlabs-testing/runs/2773438278) is with PDK 2.1.0, reproducing the original issue.

**Comment by sanfrancrisko** [comment]
Will attempt to reproduce the issue on older versions of Windows (< 2008 R2 / 8). Based on the [above comment](https://app.zenhub.com/workspaces/developer-experience-5be511b74b5806bc2bf2e38b/issues/puppetlabs/pdk/1083#issuecomment-856101703), we may need to introduce some branching to set / unset `process.leader` depending on the OS we're running on

**Comment by michaeltlombardi** [comment]
I spiked a test PR on Windows based on @sanfrancrisko's work here:
- https://github.com/puppetlabs/Puppet.Dsc/pull/168/

And it works!
- https://github.com/puppetlabs/Puppet.Dsc/pull/168/checks?check_run_id=2785882419#step:4:357

**Comment by sanfrancrisko** [comment]
We have a number of AppVeyor tests that have been failing. We got some issues resolved with #1089, however, there are still failures that seem to be specific to AppVeyor. See @jpogran 's comment [here](https://github.com/puppetlabs/pdk/pull/1089#issuecomment-859600590)

We have tested this manually on:
- Github Actions Windows Server 2019 instance
- Github Actions Windows Server 2016 instance
- ABS Windows Server 2019 instance
- ABS Windows Server 2016 instance
- ABS Windows Server 2008 R2 instance
- ABS Windows 10 instance

@michaeltlombardi also created a workflow with the config of the original issue, but this fixed pulled in: https://github.com/puppetlabs/Puppet.Dsc/pull/168/checks?check_run_id=2785882419#step:4:357
